### ISSUE TITLE
Possible Feature Request - Registration with Async Instance Creator

### ISSUE THREAD

**Issue by BrettJaner** [simpleinjector/SimpleInjector#243]
Ideally I'd like a Register overload like the following:

`Register<TService>(Func<Task<TService>> instanceCreator, Lifestyle lifestyle)`

So I could write the following:

```
container.Register<SqlConnection>(async () =>
{
     var cn = new SqlConnection("myConnString");
     await cn.OpenAsync();
     return cn;
},
Lifestyle.Scoped);
```

Would this feature make sense for Simple Injector?

**Comment by dotnetjunkie** [comment]
Mark Seemann explained that [injection constructors should be simple](http://blog.ploeh.dk/2011/03/03/InjectionConstructorsshouldbesimple/). Not only does this hold for injection constructors, but in fact this advice should be applied to the whole graph. Creation of the graph should be fast and reliable. Doing I/O not only makes it slow, but allows the graph to fail for a different reason than being wired incorrectly, and it makes it very hard to verify and diagnose the graph without the availability of this external system (in your case the database).

In fact, I would even argue that a `SqlConnection` is runtime data, and as explained [here](https://www.cuttingedge.it/blogs/steven/pivot/entry.php?id=99), runtime data should not be injected into application components. Instead the creation and especially the opening of the database connection should be postponed untill after the graph is built. This is easily done by introducing an abstraction such as an `ISqlConnectionProvider` with a `Connection` property or in your case an `Async<SqlConnection> GetConnectionAsync()` method.

**TLDR;** No, such feature does not make sense for Simple Injector or any other DI container, because it violates a few important ground rules when it comes to dependency injection.

**Comment by BrettJaner** [comment]
Yup that makes sense.  Thanks for the explanation.  I'd like to inject the same open connection into all my objects per request and have the container worry about managing the lifetime.  So would you suggest my SqlConnectionProvider implement IDisposable and look something like below?

```
    public class SqlConnectionProvider : IDisposable
    {
        private readonly Lazy<Task<SqlConnection>> _sqlConnection;

        public SqlConnectionProvider(string connectionString)
        {
            _sqlConnection = new Lazy<Task<SqlConnection>>(async () =>
            {
                var cn = new SqlConnection(connectionString);
                await cn.OpenAsync();
                return cn;
            });
        }

        public Task<SqlConnection> GetConnectionAsync()
        {
            return _sqlConnection.Value;
        }

        public void Dispose()
        {
            if (_sqlConnection.IsValueCreated)
                _sqlConnection.Value.Dispose();
        }
    }
```

**Comment by dotnetjunkie** [comment]
I think your solution is very elegant.

**Comment by BrettJaner** [comment]
Thank you :)

**Comment by dotnetjunkie** [comment]
btw, you can also let Simple Injector manage the lifetime of the `SqlConnection`; this prevents you from having to dispose it. But to be honest, I'm not sure that would make the solution more elegant. But just for completeness, here's how to do it:

``` c#
// Simplified SqlConnectionProvider implementation
public class SqlConnectionProvider {
    private readonly Func<Task<SqlConnection>> _provider;
    public SqlConnectionProvider(Func<Task<SqlConnection>> provider) { _provider = provider; }
    public Task<SqlConnection> Connection => _provider();
}
```

``` c#
// configuration
var connectionProvider = (InstanceProducer<SqlConnection>)Lifestyle.Scoped.CreateProducer(
    () => new SqlConnection(connectionString), container);

var openConnectionTaskProducer = 
    (InstanceProducer<Lazy<Task<SqlConnection>>>)Lifestyle.Scoped.CreateProducer(() =>
        new Lazy<Task<SqlConnection>>(async () => {
            var cn = connectionProvider.GetInstance();
            await cn.OpenAsync();
            return cn;
        }), container);

container.RegisterSingleton(
    new SqlConnectionProvider(() => openConnectionTaskProducer.GetInstance().Value));
```

**Comment by weitzhandler** [comment]
I wish there would have been a clear way of initializing services (not configuring or factory-ing, them), like it was mentioned earlier.

I am referring to a method that will be called after the service has been initialized, but before it's been provided.

Here's a common scenario:

```c#
AddService<TService, TImplementation>(Task<TImplementation> stuffToPerfomBeforeReturning);
```
